proposals/mutatingpolicies.md

```diff
@@ -0,0 +1,159 @@
+
+```yaml
+apiVersion: policies.kyverno.io/v1alpha1
+kind: MutatingPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == \"mesh-proxy\")"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - patchType: "ApplyConfiguration"
+      applyConfiguration:
+        expression: >
+          Object{
+            spec: Object.spec{
+              initContainers: [
+                Object.spec.initContainers{
+                  name: "mesh-proxy",
+                  image: "mesh/proxy:v1.0.0",
+                  args: ["proxy", "sidecar"],
+                  restartPolicy: "Always"
+                }
+              ]
+            }
+          }          
+```
+
+```yaml
+apiVersion: policies.kyverno.io/v1alpha1
+kind: MutatingPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == \"mesh-proxy\")"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - patchType: "JSONPatch"
+      jsonPatch:
+        expression: >
+          [
+            JSONPatch{
+              op: "add", path: "/spec/initContainers/-",
+              value: Object.spec.initContainers{
+                name: "mesh-proxy",
+                image: "mesh-proxy/v1.0.0",
+                restartPolicy: "Always"
+              }
+            }
+          ]          
+```
+
+### Mutate Existing
+
+In order to mutate existing resources, a `targets` object is added to custom existing resources:
+
+* `targets.matchConstraints`: required, selects existing resources to mutate based on GVR, `namespaceSelector`, `objectSelector`, etc.
+* `targets.matchConditions`: optional, filters target via CEL-based conditions.
+
+By default `spec.mutateExistingOnPolicyUpdate` is disabled, set it to `true` to mutate existing resources based on policy events.
+
+```yaml
+apiVersion: policies.kyverno.io/v1alpha1
+kind: MutatingPolicy
+metadata:
+  name: add-label-to-existing-secret
+spec:
+  mutateExistingOnPolicyUpdate: true
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      operations:  [CREATE]
+      resources:   ["configmaps"]
+  ## custom targets filters
+  targetMatchConstraints:
+  - # namespaceSelector:
+    # objectSelector:
+    # excludeResourceRules:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      resources:   ["secrets"]
```

Comment:
I assume we'll be using the `MatchConstraints` struct from VAPs/MAPs. The `operations` field is present, but it doesn't seem relevant for our use case. Since we won’t be generating MAPs for mutate existing rules, should we define our own struct to represent the target resources instead?